x-pack/plugins/cases/server/services/user_actions/builder.ts

```diff
@@ -0,0 +1,398 @@
+  externalService: PushedUserAction['payload']['externalService'];
+} & CommonArguments;
+type PushedBuilderResponse = BuilderReturnValue<PushedUserActionWithoutConnectorId>;
+
+// Settings
+type SettingsBuilderArgs = { settings: CaseSettings } & CommonArguments;
+type SettingsBuilderResponse = BuilderReturnValue<SettingsUserAction>;
+
+// Attachments
+type AttachmentBuilderArgs = {
+  action: UserAction;
+  comment: CommentUserAction['payload']['comment'];
+  attachmentId: string;
+} & CommonArguments;
+type AttachmentBuilderResponse = BuilderReturnValue<CommentUserAction>;
+
+// Connector
+type ConnectorBuilderArgs = {
+  connector: ConnectorUserAction['payload']['connector'];
+} & CommonArguments;
+type ConnectorBuilderResponse = BuilderReturnValue<ConnectorUserActionWithoutConnectorId>;
+
+// Create case
+type CreateCaseBuilderArgs = {
+  payload: CasePostRequest;
+} & CommonArguments;
+type CreateCaseBuilderResponse = BuilderReturnValue<CreateCaseUserActionWithoutConnectorId>;
+
+// Create case
+type DeleteCaseBuilderArgs = { connectorId: string } & CommonArguments;
+type DeleteCaseBuilderResponse = BuilderReturnValue<DeleteCaseUserAction>;
+
+type BuilderArgumentsType<T extends BuilderFieldTypes> = T extends 'title'
+  ? TitleBuilderArgs
+  : T extends 'description'
+  ? DescriptionBuilderArgs
+  : T extends 'status'
+  ? StatusBuilderArgs
+  : T extends 'tags'
+  ? TagsBuilderArgs
+  : T extends 'pushed'
+  ? PushedBuilderArgs
+  : T extends 'settings'
+  ? SettingsBuilderArgs
+  : T extends 'comment'
+  ? AttachmentBuilderArgs
+  : T extends 'connector'
+  ? ConnectorBuilderArgs
+  : T extends 'create_case'
+  ? CreateCaseBuilderArgs
+  : T extends 'delete_case'
+  ? DeleteCaseBuilderArgs
+  : never;
+
+type BuilderResponseType<T extends BuilderFieldTypes> = T extends 'title'
+  ? TitleBuilderResponse
+  : T extends 'description'
+  ? DescriptionBuilderResponse
+  : T extends 'status'
+  ? StatusBuilderResponse
+  : T extends 'tags'
+  ? TagsBuilderResponse
+  : T extends 'pushed'
+  ? PushedBuilderResponse
+  : T extends 'settings'
+  ? SettingsBuilderResponse
+  : T extends 'comment'
+  ? AttachmentBuilderResponse
+  : T extends 'connector'
+  ? ConnectorBuilderResponse
+  : T extends 'create_case'
+  ? CreateCaseBuilderResponse
+  : T extends 'delete_case'
+  ? DeleteCaseBuilderResponse
+  : never;
+
+type BuilderFunc<T extends BuilderFieldTypes> = (
+  args: BuilderArgumentsType<T>
+) => BuilderResponseType<T>;
+
+export class UserActionBuilder {
+  private readonly builderMap: Map<BuilderFieldTypes, BuilderFunc<BuilderFieldTypes>>;
+
+  private getCommonUserActionAttributes({ user, owner }: { user: User; owner: string }) {
+    return {
+      created_at: new Date().toISOString(),
+      created_by: user,
+      owner,
+    };
+  }
+
+  constructor() {
+    this.builderMap = new Map();
+    this.buildMap();
+  }
+
+  private buildMap() {
+    this.builderMap.set('title', this.buildTitleUserAction as BuilderFunc<'title'>);
```

Comment:
How do you feel about building the map dynamically so we don't have to add entries in the future?

Maybe something like:

```
    for (const field of ['title', 'comment', ...]) {
      this.builderMap.set(field, this[field] as BuilderFunc<field>);
    }
```

We'd have to name the private functions some derivative of the `field` name though.